### PR TITLE
[FLOC-4352] Set a maximum reconnect delay on the agents

### PR DESCRIPTION
**Pull request by wallrj**
Fixes: https://clusterhq.atlassian.net/browse/FLOC-4352

I couldn't figure out how to test this in a non-change-detectory way.
Perhaps we can think about it  this afternoon @sarum90 ?

Also note that latest Twisted has a new API `twisted.application.internet.ClientService` which we might be able to use here and which might make things easier to test: https://twistedmatrix.com/trac/ticket/4735

### CODE REVIEW

File: flocker/node/_loop.py [hunk 1]
```diff
@@ -53,6 +53,10 @@
 )
 from ..control._persistence import to_unserialized_json
 
+# The maximum number of seconds an agent will wait before attempting to
+# reconnect to the control service.
+MAXIMUM_RECONNECT_DELAY = 30
```
Comment (sarum90) [inline flocker/node/_loop.py]: Consider making this a command line parameter. Seems reasonable that an end-user would want to specify this value.

**Comment by sarum90** [comment]
I feel we should at least manually test this once... and maybe add a higher layer test? This is the sort of thing I'd like to have a simulator for to add a test. Although this would mean that the simulator would have to be able to fake out connection terminations, which is... interesting.

Alternatively if it a parameter you could run an acceptance test and just kill the control service for awhile?

Hrm... I might be comfortable with the tech debt of just submitting it as is. Clearly the previous behavior wasn't tested, so we aren't regressing our test coverage.

**Comment by wallrj** [comment]
> Hrm... I might be comfortable with the tech debt of just submitting it as is (after manual confirmation that this does something). Clearly the previous behavior wasn't tested, so we aren't regressing our test coverage.

@sarum90 I looked at adding a `maximium-reconnect-delay` parameter, but it looks like it'll require various refactoring and changes to tests.

So for now I'm going to leave this as-is and hopefully the following log trace is sufficient evidence that the agent reconnect delay never gets beyond 30s (+/- jitter)

```
(4352) (fast-reconnect-FLOC-4352 ? =)[~/.../HybridLogic/flocker]$ eliot-prettyprint < flocker-dataset-agent.log 
2a3819d5-e87a-4aaf-99ce-58fdd7b57eb0 -> /1
2016-04-29 22:29:26.780955Z
  message_type: 'twisted:log'
  error: False
  message: 'Log opened.'

90d9248b-856c-4bcf-9e83-c37fda65f638 -> /1
2016-04-29 22:29:27.525130Z
  action_type: 'flocker:node:script:get_external_ip'
  action_status: 'started'
  host: '127.0.0.1'
  port: 4524

90d9248b-856c-4bcf-9e83-c37fda65f638 -> /2
2016-04-29 22:29:27.526224Z
  action_type: 'flocker:node:script:get_external_ip'
  action_status: 'succeeded'
  local_ip: '127.0.0.1'

84a7f911-25d1-4ed3-af25-258086a7eeef -> /1
2016-04-29 22:29:27.542156Z
  action_type: 'fsm:initialize'
  action_status: 'started'
  fsm_identifier: '<flocker.node._loop.ConvergenceLoop object at 0x7fd82ab59d90>'
  fsm_state: '<ConvergenceLoopStates=STOPPED>'

1f2dc36a-d096-4f3e-9028-ff1f1e3055a5 -> /1
2016-04-29 22:29:27.543163Z
  action_type: 'fsm:initialize'
  action_status: 'started'
  fsm_identifier: '<flocker.node._loop.ClusterStatus object at 0x7fd82aaf9050>'
  fsm_state: '<ClusterStatusStates=DISCONNECTED>'

6bcbe9ff-da6f-4f92-a55d-a47827dd969a -> /1
2016-04-29 22:29:27.544162Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

ca8c9393-fe95-49f8-ba93-b4f28326be1f -> /1
2016-04-29 22:29:27.546005Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 3 seconds'

a6c42cb6-dbaf-4416-a1c0-1b8b662250cf -> /1
2016-04-29 22:29:27.546613Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

6406a70b-2af0-47d9-9871-5065ea4a5bb1 -> /1
2016-04-29 22:29:31.105862Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

2904baf1-f594-4022-810b-6ca222c7faed -> /1
2016-04-29 22:29:31.106668Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 7 seconds'

30d27771-4365-44cf-a3bd-6fcbe01c1603 -> /1
2016-04-29 22:29:31.106988Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

a120d9d6-edc7-4a25-9da8-8aa21c32ff95 -> /1
2016-04-29 22:29:38.994880Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

feb6d2f3-2d68-44b5-8746-d8e8a20fb35d -> /1
2016-04-29 22:29:38.996778Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 23 seconds'

dca22c54-a123-4260-9756-a2d3167b5663 -> /1
2016-04-29 22:29:38.997508Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

34a21ceb-70b6-442a-aa77-758dc218a623 -> /1
2016-04-29 22:30:02.237584Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

e38252b8-3b87-48e0-966c-a5e176e04abc -> /1
2016-04-29 22:30:02.239404Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 31 seconds'

883ac2eb-22e0-4ffc-8584-5a05d070af41 -> /1
2016-04-29 22:30:02.240243Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

45f3bf74-8282-496e-88a8-8848c6bfd469 -> /1
2016-04-29 22:30:33.395847Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

19e6f58a-2131-4828-97f5-329aeee7dbc6 -> /1
2016-04-29 22:30:33.397662Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 32 seconds'

224da612-4026-465d-b6e4-7c57349006f0 -> /1
2016-04-29 22:30:33.398274Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

ef02cd13-c787-4736-b564-fffe96d243dc -> /1
2016-04-29 22:31:06.128562Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

b9f23e3a-551d-49d7-8a15-4c4745fc8622 -> /1
2016-04-29 22:31:06.129477Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 34 seconds'

dddfcc8a-73a7-4943-ac61-6dd383696bc0 -> /1
2016-04-29 22:31:06.129805Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

f66a5fda-514f-4ecf-8b43-d8357cea4b88 -> /1
2016-04-29 22:31:40.497023Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

2d160351-5e2d-4ef2-887d-d3951767889c -> /1
2016-04-29 22:31:40.498850Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 29 seconds'

e1775427-d6ac-4601-82a8-89a924fbb261 -> /1
2016-04-29 22:31:40.499799Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

adc6fe9b-d34f-4d97-981e-a2763068291c -> /1
2016-04-29 22:32:10.487138Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

ffabee11-5f54-4f72-8799-82340e0a5b9e -> /1
2016-04-29 22:32:10.489504Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 27 seconds'

465e187a-3970-449c-b28a-79a59bb2f63a -> /1
2016-04-29 22:32:10.490712Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

e687f20a-c942-4574-b7ca-f49e319fb3c7 -> /1
2016-04-29 22:32:38.328036Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

12cb5d64-af9d-46a1-8c75-84815a0f383d -> /1
2016-04-29 22:32:38.329708Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 29 seconds'

544030d4-cfc5-43ea-a7f6-db3224370db2 -> /1
2016-04-29 22:32:38.330534Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

dea1d905-318a-4d9d-bf47-93d7c45bf551 -> /1
2016-04-29 22:33:07.516012Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

8120f59d-034c-4401-8d43-5a0b74e85dde -> /1
2016-04-29 22:33:07.519543Z
  message_type: 'twisted:log'
  error: False
  message: "AgentAMP connection established (HOST:IPv4Address(TCP, '127.0.0.1', 47808) PEER:IPv4Address(TCP, '127.0.0.1', 4524))"

97f00ed7-ebc7-4b31-8091-a3cc1b9d0365 -> /1
2016-04-29 22:33:07.520555Z
  action_type: 'fsm:transition'
  action_status: 'started'
  fsm_identifier: '<flocker.node._loop.ClusterStatus object at 0x7fd82aaf9050>'
  fsm_input: '<ClusterStatusInputs=CONNECTED_TO_CONTROL_SERVICE>'
  fsm_rich_input: '<_ConnectedToControlService>'
  fsm_state: '<ClusterStatusStates=DISCONNECTED>'

97f00ed7-ebc7-4b31-8091-a3cc1b9d0365 -> /2
2016-04-29 22:33:07.520885Z
  action_type: 'fsm:transition'
  action_status: 'succeeded'
  fsm_next_state: '<ClusterStatusStates=IGNORANT>'
  fsm_output: ['<ClusterStatusOutputs=STORE_CLIENT>']

1f5b9ff3-aa27-4b40-928d-749ef4b22f55 -> /1
2016-04-29 22:34:07.547437Z
  message_type: 'twisted:log'
  error: False
  message: "AgentAMP connection lost (HOST:IPv4Address(TCP, '127.0.0.1', 47808) PEER:IPv4Address(TCP, '127.0.0.1', 4524))"

cc1322d7-18b3-48e0-bb57-566c3ff60139 -> /1
2016-04-29 22:34:07.553147Z
  message_type: 'eliot:traceback'
  exception: 'OpenSSL.SSL.Error'
  reason: "[('SSL routines', 'ssl23_read', 'ssl handshake failure')]"
  traceback: "Traceback: <class 'OpenSSL.SSL.Error'>: [('SSL routines', 'ssl23_read', 'ssl handshake failure')]
           |  /home/richard/.virtualenvs/4352/lib/python2.7/site-packages/twisted/internet/base.py:825:runUntilCurrent
           |  /home/richard/.virtualenvs/4352/lib/python2.7/site-packages/twisted/internet/tcp.py:479:connectionLost
           |  /home/richard/.virtualenvs/4352/lib/python2.7/site-packages/twisted/internet/tcp.py:293:connectionLost
           |  /home/richard/.virtualenvs/4352/lib/python2.7/site-packages/twisted/protocols/tls.py:473:connectionLost
           |  --- <exception caught here> ---
           |  /home/richard/.virtualenvs/4352/lib/python2.7/site-packages/twisted/protocols/tls.py:360:_flushReceiveBIO
           |  /home/richard/.virtualenvs/4352/lib/python2.7/site-packages/OpenSSL/SSL.py:1320:recv
           |  /home/richard/.virtualenvs/4352/lib/python2.7/site-packages/OpenSSL/SSL.py:1187:_raise_ssl_error
           |  /home/richard/.virtualenvs/4352/lib/python2.7/site-packages/OpenSSL/_util.py:48:exception_from_error_queue
           |  "

9e943702-ce78-4e0b-aa3a-8fc43c9ced2b -> /1
2016-04-29 22:34:07.554550Z
  action_type: 'fsm:transition'
  action_status: 'started'
  fsm_identifier: '<flocker.node._loop.ClusterStatus object at 0x7fd82aaf9050>'
  fsm_input: '<ClusterStatusInputs=DISCONNECTED_FROM_CONTROL_SERVICE>'
  fsm_rich_input: None
  fsm_state: '<ClusterStatusStates=IGNORANT>'

9e943702-ce78-4e0b-aa3a-8fc43c9ced2b -> /2
2016-04-29 22:34:07.555299Z
  action_type: 'fsm:transition'
  action_status: 'succeeded'
  fsm_next_state: '<ClusterStatusStates=DISCONNECTED>'
  fsm_output: []

77be0019-c81e-414a-8400-bda0e3020d87 -> /1
2016-04-29 22:34:07.556068Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 2 seconds'

a468c87e-a0e2-461c-81c3-0fb6859cd216 -> /1
2016-04-29 22:34:07.556782Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

a9b6f5b6-ff05-4efe-b067-00dc614d6fa7 -> /1
2016-04-29 22:34:09.687077Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

f0dec5b8-626e-45e4-ac39-2283015c0f9f -> /1
2016-04-29 22:34:09.688366Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 5 seconds'

1c3385f4-48a1-4355-82d1-925c26921250 -> /1
2016-04-29 22:34:09.688853Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

f02b6391-db5f-438f-b061-fc4ced243c0a -> /1
2016-04-29 22:34:15.608774Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

fd3012f3-dc4e-47af-a6bd-14d07f377ba9 -> /1
2016-04-29 22:34:15.609787Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 18 seconds'

91252e83-7408-4ea9-995e-0d7003bc59d5 -> /1
2016-04-29 22:34:15.610148Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

8cc604f9-1b49-4baf-9a8a-996481ad25b2 -> /1
2016-04-29 22:34:34.088230Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

13568b9e-648b-4fa3-b248-c8457bf2bed7 -> /1
2016-04-29 22:34:34.089235Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 34 seconds'

5ce3fd6a-c04b-42d9-b9d7-cbf13cd2560c -> /1
2016-04-29 22:34:34.089596Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

7ab6838a-e922-4897-b800-9e08b9b7eff3 -> /1
2016-04-29 22:35:09.066717Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

49766739-5e9c-47b6-9daf-89dfb0ed9b82 -> /1
2016-04-29 22:35:09.068284Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 23 seconds'

ffacbe27-518e-4c6c-9268-385bc3a92d36 -> /1
2016-04-29 22:35:09.069108Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

ae2c4836-e14a-4988-99f1-c015bd140c66 -> /1
2016-04-29 22:35:32.204781Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

bcd1bfea-2322-42c4-b909-3347c60713d7 -> /1
2016-04-29 22:35:32.205719Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 31 seconds'

58beeac5-51d9-4d81-b3eb-6b13246c6e89 -> /1
2016-04-29 22:35:32.206239Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

0480c8e0-2ede-4638-addd-ab512deaa93a -> /1
2016-04-29 22:36:03.403154Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

7b994c9e-8715-487d-ad5a-cd3795ed3afe -> /1
2016-04-29 22:36:03.404534Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 23 seconds'

dfbca6f4-9f50-4321-a6b9-18e0a0c004a0 -> /1
2016-04-29 22:36:03.405018Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

7b33e6a6-de33-45e4-ad1e-af04abdbbd49 -> /1
2016-04-29 22:36:26.942398Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

0d778b83-129f-4d4b-abbc-bc6651871faa -> /1
2016-04-29 22:36:26.943786Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 29 seconds'

02042617-5d20-4e63-95e0-ef938b724a16 -> /1
2016-04-29 22:36:26.944317Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

de618491-e373-4e3a-aee8-c9914c61714a -> /1
2016-04-29 22:36:56.287928Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

7c37636c-3908-4014-8ffd-4be705918ff6 -> /1
2016-04-29 22:36:56.289445Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 30 seconds'

02fff0aa-60f9-45f9-b3cf-04e8e57091c9 -> /1
2016-04-29 22:36:56.290020Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

a5a9b357-d884-4b37-a7ab-372b6cbb7b05 -> /1
2016-04-29 22:37:26.384057Z
  message_type: 'twisted:log'
  error: False
  message: 'Starting factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'

839b7c27-8a0e-4952-96a8-ba6de10968a4 -> /1
2016-04-29 22:37:26.385975Z
  message_type: 'twisted:log'
  error: False
  message: '<twisted.internet.tcp.Connector instance at 0x7fd82b5c6b90> will retry in 30 seconds'

8da10c34-7492-4505-a2b4-e0686f8ab1bd -> /1
2016-04-29 22:37:26.386820Z
  message_type: 'twisted:log'
  error: False
  message: 'Stopping factory <twisted.internet.protocol.ReconnectingClientFactory instance at 0x7fd82aaf6488>'


```

**Comment by wallrj** [comment]
Here's what's required to implement the `maximum-reconnect-delay`:

```#!diff
diff --git a/flocker/node/_loop.py b/flocker/node/_loop.py
index e7603e6..6d41e5e 100644
--- a/flocker/node/_loop.py
+++ b/flocker/node/_loop.py
@@ -728,7 +728,8 @@ class RemoteStatePersister(PClass):
 
 
 @implementer(IConvergenceAgent)
-@attributes(["reactor", "deployer", "host", "port", "era"])
+@attributes(["reactor", "deployer", "host", "port", "era",
+             "maximum_reconnect_delay"])
 class AgentLoopService(MultiService, object):
     """
     Service in charge of running the convergence loop.
@@ -758,7 +759,7 @@ class AgentLoopService(MultiService, object):
         self.reconnecting_factory = ReconnectingClientFactory.forProtocol(
             lambda: AgentAMP(self.reactor, self)
         )
-        self.reconnecting_factory.maxDelay = MAXIMUM_RECONNECT_DELAY
+        self.reconnecting_factory.maxDelay = self.maximum_reconnect_delay
         self.factory = TLSMemoryBIOFactory(context_factory, True,
                                            self.reconnecting_factory)
 
diff --git a/flocker/node/script.py b/flocker/node/script.py
index 2f48e14..f29e557 100644
--- a/flocker/node/script.py
+++ b/flocker/node/script.py
@@ -248,6 +248,10 @@ class _AgentOptions(Options):
     optParameters = [
         ["agent-config", "c", "/etc/flocker/agent.yml",
          "The configuration file to set the node service."],
+        ["maximum-reconnect-delay", None, 30,
+         "The maximum number of seconds to wait "
+         "before attempting to connect to the control service "
+         "after a failed connection attempt.", int],
     ]
 
     def postOptions(self):
@@ -339,13 +343,14 @@ class AgentServiceFactory(PClass):
 
         tls_info = _context_factory_and_credential(
             options["agent-config"].parent(), host, port)
-
         return AgentLoopService(
             reactor=reactor,
             deployer=self.deployer_factory(
                 node_uuid=tls_info.node_credential.uuid, hostname=ip,
                 cluster_uuid=tls_info.node_credential.cluster_uuid),
-            host=host, port=port,
+            host=host,
+            port=port,
+            maximum_reconnect_delay=options['maximum-reconnect-delay'],
             context_factory=tls_info.context_factory,
             era=get_era(),
         )
@@ -554,7 +559,7 @@ class AgentService(PClass):
             api=api, hostname=address, node_uuid=node_uuid,
         )
 
-    def get_loop_service(self, deployer):
+    def get_loop_service(self, deployer, options):
         """
         :param IDeployer deployer: The deployer which the loop service can use
             to interact with the system.
@@ -569,6 +574,7 @@ class AgentService(PClass):
             host=self.control_service_host, port=self.control_service_port,
             context_factory=self.get_tls_context().context_factory,
             era=get_era(),
+            maximum_reconnect_delay=options["maximum-reconnect-delay"],
         )
 
 
@@ -595,7 +601,7 @@ class DatasetServiceFactory(PClass):
 
         deployer = agent_service.get_deployer(api)
 
-        loop_service = agent_service.get_loop_service(deployer)
+        loop_service = agent_service.get_loop_service(deployer, options)
 
         return loop_service
 
```

**Pull request merged**